### PR TITLE
RDP classifier

### PR DESCRIPTION
**Pull request by taylorpaisie**
Pull Request (PR) checklist:
- [x ] Include a description of what is in this pull request in this message.
- [ x] The dockerfile successfully builds to a test target for the user creating the PR. (i.e. `docker build --tag samtools:1.15test --target test docker-builds/samtools/1.15` )
- [ x] Directory structure as name of the tool in lower case with special characters removed with a subdirectory of the version number (i.e. `spades/3.12.0/Dockerfile`)
   - [ ] (optional) All test files are located in same directory as the Dockerfile (i.e. `shigatyper/2.0.1/test.sh`)
- [ x] Create a simple container-specific [README.md](https://github.com/StaPH-B/docker-builds/blob/master/.github/workflow-templates/readme-template.md) in the same directory as the Dockerfile (i.e. `spades/3.12.0/README.md`)
   - [ x] If this README is longer than 30 lines, there is an explanation as to why more detail was needed
- [ x] Dockerfile includes the recommended [LABELS](https://github.com/StaPH-B/docker-builds/blob/master/dockerfile-template/Dockerfile#L8-L18) 
- [ x] Main [README.md](https://github.com/StaPH-B/docker-builds/blob/master/README.md) has been updated to include the tool and/or version of the dockerfile(s) in this PR
- [ x] [Program_Licenses.md](https://github.com/StaPH-B/docker-builds/blob/master/Program_Licenses.md) contains the tool(s) used in this PR and has been updated for any missing

### CODE REVIEW

**Comment by chrisgulvik** [comment]
FYI this is a super fast and useful Genus-level 16S rRNA gene classifier tool. Part of the reason the container is bigger than most is it includes the database file within the package. There's no documentation to change/create the database either. The developer left MSU (RDP Project) years ago for industry, and just made a code update (with new builtin DB) last year on her own time.

**Comment by erinyoung** [comment]
This image looks great! Thank you for putting it together. 

The GA failed because your labels are in the wrong stage. Could you move them to 'app'?

Also, is there are reason why you are using so many `ENV` variables? `ENV` persist after build time, so we generally recommend to use `ARG` instead so that the environment doesn't end up with unused environmental variables.

**Comment by erinyoung** [comment]
Your tests all passed:
```
#17 [test 4/4] RUN python3 -m unittest discover -v -s ../tests
#17 0.390 2024-08-28 20:06:06 URL:https://raw.githubusercontent.com/taylorpaisie/docker_containers/main/rdp/2.14/16S_rRNA_gene.Burkholderia_pseudomallei.2002721184.AY305776.1.fasta [1629/1629] -> "16S_test.fa" [1]
#17 0.613 2024-08-28 20:06:06 URL:https://raw.githubusercontent.com/taylorpaisie/docker_containers/main/rdp/2.14/18S_rRNA_gene.Homo_sapiens.T2T-CHM13v2.0.Chromosome13.fasta [2026/2026] -> "18S_test.fa" [1]
#17 0.687 Exception in thread "main" java.lang.InternalError: Error loading java.security file
#17 0.687 	at java.base/java.security.Security.initialize(Security.java:94)
#17 0.688 	at java.base/java.security.Security$1.run(Security.java:79)
#17 0.688 	at java.base/java.security.Security$1.run(Security.java:77)
#17 0.688 	at java.base/java.security.AccessController.doPrivileged(Native Method)
#17 0.688 	at java.base/java.security.Security.<clinit>(Security.java:77)
#17 0.688 	at java.base/sun.security.util.SecurityProperties.getOverridableProperty(SecurityProperties.java:53)
#17 0.688 	at java.base/sun.security.util.SecurityProperties.privilegedGetOverridable(SecurityProperties.java:44)
#17 0.688 	at java.base/sun.security.util.FilePermCompat.<clinit>(FilePermCompat.java:45)
#17 0.688 	at java.base/java.io.FilePermission.init(FilePermission.java:317)
#17 0.688 	at java.base/java.io.FilePermission.<init>(FilePermission.java:488)
#17 0.688 	at java.base/sun.net.www.protocol.file.FileURLConnection.getPermission(FileURLConnection.java:221)
#17 0.688 	at java.base/sun.net.www.protocol.jar.JarFileFactory.getPermission(JarFileFactory.java:162)
#17 0.688 	at java.base/sun.net.www.protocol.jar.JarFileFactory.getCachedJarFile(JarFileFactory.java:126)
#17 0.689 	at java.base/sun.net.www.protocol.jar.JarFileFactory.get(JarFileFactory.java:81)
#17 0.689 	at java.base/sun.net.www.protocol.jar.JarURLConnection.connect(JarURLConnection.java:125)
#17 0.689 	at java.base/sun.net.www.protocol.jar.JarURLConnection.getInputStream(JarURLConnection.java:155)
#17 0.689 	at java.base/java.net.URL.openStream(URL.java:1165)
#17 0.689 	at java.base/java.lang.ClassLoader.getResourceAsStream(ClassLoader.java:1744)
#17 0.689 	at java.base/java.lang.Class.getResourceAsStream(Class.java:2650)
#17 0.689 	at edu.msu.cme.rdp.classifier.utils.ClassifierFactory.getFactory(ClassifierFactory.java:110)
#17 0.689 	at edu.msu.cme.rdp.multicompare.MultiClassifier.<init>(MultiClassifier.java:64)
#17 0.689 	at edu.msu.cme.rdp.multicompare.MultiClassifier.<init>(MultiClassifier.java:75)
#17 0.689 	at edu.msu.cme.rdp.multicompare.Main.main(Main.java:247)
#17 0.689 	at edu.msu.cme.rdp.classifier.cli.ClassifierMain.main(ClassifierMain.java:67)
#17 0.781 Exception in thread "main" java.lang.InternalError: Error loading java.security file
#17 0.781 	at java.base/java.security.Security.initialize(Security.java:94)
#17 0.781 	at java.base/java.security.Security$1.run(Security.java:79)
#17 0.781 	at java.base/java.security.Security$1.run(Security.java:77)
#17 0.782 	at java.base/java.security.AccessController.doPrivileged(Native Method)
#17 0.782 	at java.base/java.security.Security.<clinit>(Security.java:77)
#17 0.782 	at java.base/sun.security.util.SecurityProperties.getOverridableProperty(SecurityProperties.java:53)
#17 0.782 	at java.base/sun.security.util.SecurityProperties.privilegedGetOverridable(SecurityProperties.java:44)
#17 0.782 	at java.base/sun.security.util.FilePermCompat.<clinit>(FilePermCompat.java:45)
#17 0.782 	at java.base/java.io.FilePermission.init(FilePermission.java:317)
#17 0.782 	at java.base/java.io.FilePermission.<init>(FilePermission.java:488)
#17 0.782 	at java.base/sun.net.www.protocol.file.FileURLConnection.getPermission(FileURLConnection.java:221)
#17 0.782 	at java.base/sun.net.www.protocol.jar.JarFileFactory.getPermission(JarFileFactory.java:162)
#17 0.782 	at java.base/sun.net.www.protocol.jar.JarFileFactory.getCachedJarFile(JarFileFactory.java:126)
#17 0.782 	at java.base/sun.net.www.protocol.jar.JarFileFactory.get(JarFileFactory.java:81)
#17 0.782 	at java.base/sun.net.www.protocol.jar.JarURLConnection.connect(JarURLConnection.java:125)
#17 0.782 	at java.base/sun.net.www.protocol.jar.JarURLConnection.getInputStream(JarURLConnection.java:155)
#17 0.782 	at java.base/java.net.URL.openStream(URL.java:1165)
#17 0.783 	at java.base/java.lang.ClassLoader.getResourceAsStream(ClassLoader.java:1744)
#17 0.783 	at java.base/java.lang.Class.getResourceAsStream(Class.java:2650)
#17 0.783 	at edu.msu.cme.rdp.classifier.utils.ClassifierFactory.getFactory(ClassifierFactory.java:110)
#17 0.783 	at edu.msu.cme.rdp.multicompare.MultiClassifier.<init>(MultiClassifier.java:64)
#17 0.783 	at edu.msu.cme.rdp.multicompare.MultiClassifier.<init>(MultiClassifier.java:75)
#17 0.783 	at edu.msu.cme.rdp.multicompare.Main.main(Main.java:247)
#17 0.783 	at edu.msu.cme.rdp.classifier.cli.ClassifierMain.main(ClassifierMain.java:67)
#17 0.791 test_rdp16S (test_controls.TestControls) ... ok
#17 0.791 test_rdp18S (test_controls.TestControls) ... ok
#17 0.791 test_python (test_versions.TestVersion) ... ok
#17 0.792 
#17 0.792 ----------------------------------------------------------------------
#17 0.792 Ran 3 tests in 0.680s
#17 0.792 
#17 0.792 OK
#17 DONE 0.8s
```

File: rdp/2.14/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    curl \
+    unzip 
+
+ENV JAVA_HOME /usr/lib/jvm/java-11-openjdk-amd64
+ENV PATH="${PATH}:${JAVA_HOME}/bin"
+
+# Install RDP
+ENV RDP_ZIP rdp_classifier_2.14.zip
+ENV RDP_DOWNLOAD https://sourceforge.net/projects/rdp-classifier/files/rdp-classifier/rdp_classifier_2.14.zip/download
+ENV RDP_PATH /rdp_classifier_2.14
```
Comment (erinyoung) [inline rdp/2.14/Dockerfile:33]: Are these environmental variables needed when running rpd?

File: rdp/2.14/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    wget \
+    curl \
+    unzip 
+
+ENV JAVA_HOME /usr/lib/jvm/java-11-openjdk-amd64
+ENV PATH="${PATH}:${JAVA_HOME}/bin"
+
+# Install RDP
+ENV RDP_ZIP rdp_classifier_2.14.zip
+ENV RDP_DOWNLOAD https://sourceforge.net/projects/rdp-classifier/files/rdp-classifier/rdp_classifier_2.14.zip/download
```
Comment (erinyoung) [inline rdp/2.14/Dockerfile:32]: Could you replace the hardcoded version with a variable? This makes it easier to update these dockerfiles later.

Something like...
```
ARG RDP_ZIP=rdp_classifier_${RDP_VER}.zip
ARG RDP_DOWNLOAD https://sourceforge.net/projects/rdp-classifier/files/rdp-classifier/rdp_classifier_${RDP_VER}.zip/download
```

File: rdp/2.14/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+RUN wget $RDP_DOWNLOAD -O /tmp/$RDP_ZIP &&\
+  unzip /tmp/$RDP_ZIP -d /tmp &&\
+  mv /tmp/${RDP_ZIP%.*} $RDP_PATH &&\
+  rm -rf /tmp/$RDP_ZIP &&\
+  chmod +x ${RDP_PATH}/dist/classifier.jar &&\
+  echo "#!/bin/bash" >> /${RDP_PATH}/dist/classifier &&\
+  echo "exec java -jar /${RDP_PATH}/dist/classifier.jar """"$""@"""" " >> /${RDP_PATH}/dist/classifier &&\
+  chmod +x /${RDP_PATH}/dist/classifier
+
+ENV PATH="${PATH}:/${RDP_PATH}/dist/"
```
Comment (erinyoung) [inline rdp/2.14/Dockerfile:44]: I wouldn't worry about defining PATH in builder, so this line can be removed

File: rdp/2.14/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+
+# Running the dockerfile and tests
+FROM ubuntu:focal as app
+ARG RDP_VER
+
+COPY --from=builder /rdp_classifier_2.14/ /rdp_classifier_2.14/
```
Comment (erinyoung) [inline rdp/2.14/Dockerfile:51]: It might make things easier if the version were removed from this directory

File: rdp/2.14/README.md [hunk 5]
```diff
@@ -0,0 +1,29 @@
+# RDP Classifier 
+
+This image implements:
+* [RDP v2.14](https://sourceforge.net/projects/rdp-classifier/)
+
+and can be accessed at [docker hub](https://hub.docker.com/u/tpaisie).
+
```
Comment (erinyoung) [inline rdp/2.14/README.md:7]: Can you add in some additional information about how to use rdp?

```
Main tool: [RDP Classifier](https://sourceforge.net/projects/rdp-classifier/)
  
Code repository: https://github.com/rdpstaff/classifier

Basic information on how to use this tool:
- executable: |
	classify      - classify one or multiple samples
	crossvalidate - cross validate accuracy testing
	libcompare    - compare two samples
	loot          - leave one (sequence or taxon) out accuracy testing
	merge-detail  - merge classification detail result files to create a taxon assignment counts file
	merge-count   - merge multiple taxon assignment count files to into one count file
	random-sample - random select a subset or subregion of sequences
	rm-dupseq     - remove identical or any sequence contained by another sequence
	rm-partialseq - remove partial sequences
	taxa-sim      - calculate and plot the similarities within taxa
	train         - retrain classifier

- help: NA
- version: NA
- description: | 
> The RDP Classifier is a naive Bayesian classifier which was developed to provide rapid taxonomic placement based on rRNA sequence data.

Additional information:

<Container contains X database at Y>
  
Full documentation: link to documentation or wiki : https://sourceforge.net/projects/rdp-classifier/
```

**Comment by erinyoung** [comment]
I'm trying to provide some guidance on a CMD layer. This would be the default command that runs when given no arguments (i.e. docker run staphb/rdp:latest). Generally we use <tool --help>, but that doesn't seem to work for rdp.

**Comment by chrisgulvik** [comment]
> I'm trying to provide some guidance on a CMD layer. This would be the default command that runs when given no arguments (i.e. docker run staphb/rdp:latest). Generally we use <tool --help>, but that doesn't seem to work for rdp.

@erinyoung Only `rdp_classifier` (no args) brings up the help menu with the subcommand options (e.g., `rdp_classifier classify`). Both `-h` and `--help` still print the help menu including a 1-line warning that neither is a supported flag, although both return 0 exit status still.

**Comment by erinyoung** [comment]
Sweet! Thank you for your response! Can I push some edits your dockerfile? I'd like to move the labels, change some ENV to ARG, and add a CMD line.

**Comment by taylorpaisie** [comment]
Yeah go for it @erinyoung, sorry been meaning to make some corrections, just haven't had the time yet! But thank you so much for your help!

**Comment by erinyoung** [comment]
@taylorpaisie , as I was looking over your Dockerfile, it doesn't look like this tool requires compiling, so the 'builder' stage may not be worth the effort.

I've made some adjustments to your Dockerfile and README.md in this PR. Let me know if you think these will work.

For the Dockerfile: 
Removed most of the 'ENV' layers except for the PATH update
Removed the builder stage and instead reduces the amount downloaded with apt-get
Switched hard-coded filepath to one with an ARG (it's easier to update these in the future)
Switched the base.image to jammy instead of focal (jammy is newer, so the image will have a longer lifespan)
Adjusted the path of the repository
Added CMD layer (this is the command the executes if no command is given to the image)
Added a new workdir for the test stage
Added classifier to test stage (to ensure it was in path)
Updated python version test to 3.10

The only things left are to add rdp/classifier to the following two files:
- README.md at root (https://github.com/taylorpaisie/docker-builds/blob/tkp-rdp/README.md)
- Program_licenses.md at root (https://github.com/taylorpaisie/docker-builds/blob/tkp-rdp/Program_Licenses.md)

I would also encourage you (and anyone else who assisted with this) to add your github handle to the bottom of the README.md at root. I love it when people get credit for contributing work.

File: README.md [hunk 6]
```diff
@@ -251,6 +251,7 @@ To learn more about the docker pull rate limits and the open source software pro
 | [raven](https://hub.docker.com/r/staphb/raven/) <br/> [![docker pulls](https://badgen.net/docker/pulls/staphb/raven)](https://hub.docker.com/r/staphb/raven) | <ul><li>1.5.1</li><li>1.8.1</li><li>[1.8.3](./raven/1.8.3)</li></ul> | https://github.com/lbcb-sci/raven |
 | [RAxML](https://hub.docker.com/r/staphb/raxml/) <br/> [![docker pulls](https://badgen.net/docker/pulls/staphb/raxml)](https://hub.docker.com/r/staphb/raxml) | <ul><li>8.2.12</li><li>[8.2.13](./raxml/8.2.13/)</li></ul> | https://github.com/stamatak/standard-RAxML |
 | [RAxML-NG](https://hub.docker.com/r/staphb/raxml-ng/) <br/> [![docker pulls](https://badgen.net/docker/pulls/staphb/raxml-ng)](https://hub.docker.com/r/staphb/raxml-ng) | <ul><li>[1.2.2](./raxml-ng/1.2.2/)</li></ul> | https://github.com/amkozlov/raxml-ng |
+| [rdp](https://hub.docker.com/r/staphb/rdp) <br/> [![docker pulls](https://badgen.net/docker/pulls/staphb/rdp)](https://hub.docker.com/r/staphb/rdp) | <ul><li>[2.14](./rdp/0.4.0/)</li></ul> | https://sourceforge.net/projects/rdp-classifier/files/rdp-classifier/rdp_classifier_2.14.zip/download |
```
Comment (erinyoung) [inline README.md:254]: Did you want to add your user handle to the bottom of this file? (Optional)

Comment (taylorpaisie) [reply README.md:254]: added!

**Review by erinyoung — APPROVED** [review]
I have no additional changes to recommend. I approve this PR. 

Once the tests complete (again), I'll merge this PR and deploy the image that this Dockerfile creates.

**Comment by erinyoung** [comment]
Thank you for putting this together!

The GA that are using your Dockerfile to build an image and deploy it to Dockerhub and quay can be found here : https://github.com/StaPH-B/docker-builds/actions/runs/10801924836

Once this action is complete, the image should be available at https://hub.docker.com/r/staphb/rdp with the tags of 'latest' and '2.14'

**Pull request merged**